### ISSUE TITLE
`twistd conch´ always fails due to an unhandled AttributeError; conch server cannot be started.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3871]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3871|
|Type|release blocker: regression|
|Created|2009-06-06 12:51:05Z|
|Branch|https://github.com/twisted/twisted/tree/conch-tap-fix-3871-2|

```
exarkun@boson:~$ twistd -n conch
/home/exarkun/Projects/Divmod/trunk/Nevow/nevow/static.py:35: DeprecationWarning: twisted.web.error.NoResource is deprecated since Twisted 9.0.  See twisted.web.resource.NoResource.
  dangerousPathError = error.NoResource("Invalid request URL.")
Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/bin/twistd", line 19, in <module>
    run()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/scripts/twistd.py", line 27, in run
    app.run(runApp, ServerOptions)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/application/app.py", line 690, in run
    runApp(config)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/scripts/twistd.py", line 23, in runApp
    _SomeApplicationRunner(config).run()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/application/app.py", line 411, in run
    self.application = self.createOrGetApplication()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/application/app.py", line 489, in createOrGetApplication
    ser = plg.makeService(self.config.subOptions)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/conch/tap.py", line 40, in makeService
    t.portal.registerChecker(checkers.PluggableAuthenticationModulesChecker())
AttributeError: 'module' object has no attribute 'PluggableAuthenticationModulesChecker'
exarkun@boson:~$ 
```

It seems most likely that this was introduced by [#2682](https://github.com/twisted/twisted/issues/2682) (although I don't believe PAM-based authentication was working prior to that either, however the failure mode was silent afaict).

Attachments:

* [fix](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4fd/4fd67f03c9511028bb713318efc0e01917864e38/99f707ef02f096ed1e08d922e18da99b31d43e4f) (1552 bytes) - added by exarkun on 2009-06-06 12:55:44Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3871 3871
type__release_blocker__regression release blocker: regression
reporter__exarkun exarkun
priority__highest highest
milestone__Twisted_9_0 Twisted-9.0
branch__branches_conch_tap_fix_3871_2 branches/conch-tap-fix-3871-2
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1244292665000000 1244292665000000
changetime__1246950306000000 1246950306000000
version__None None
owner__ 
cc__glyph cc__mwh
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

This seems to be what happened.

  * Ages, ago, `PluggableAuthenticationModulesChecker` moved out of `twisted.conch.checkers` and into `twisted.cred.checkers`.
  * Along with it, `twisted.conch.pamauth` moved into `twisted.cred`.
  * At that point, `twisted.conch.tap` became wrong, as it continued to think these things were in `twisted.conch`.
  * However, the failure was silent, since it manifest as the plugin always detecting PAM as unavailable (since it continued to import `twisted.conch.pamauth` via `twisted.conch.checkers`).  It handled the case of PAM being missing by not setting up a PAM checker.
  * [#2682](https://github.com/twisted/twisted/issues/2682) changed `twisted.conch.checkers` to import `twisted.cred.pamauth` instead of `twisted.conch.pamauth`.  This seems to have been a change with no unit test.  It was also pointless, since nothing in `twisted.conch.checkers` _uses_ `pamauth`.  And of course, it was also detrimental, as it made a silent failure that wasn't _fatal_ into a fatal failure that prevents the server from being run at all.

Attaching a patch which demonstrates one way this can be remedied.  Of course it is not suitable for inclusion as-is, as it has no test coverage.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Ah, and as this is a regression from 8.2, we really ought to fix it for 9.0.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [27034]) Branching to 'conch-tap-fix-3871'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

Please review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @therve<br>@exarkun set status to `new`|
|-|-|

Thanks for working on this.

  1. `twisted/conch/tap.py` could have a `test-case-name` now
  1. `test_makeServicePamAuth` should use a set or something when comparing `portal.checkers.keys()`, to avoid relying on dict ordering.
  1. It would be cool to have a non-skipped test for the without-pamauth case.  Right now if I break the implementation and uninstall python-pam, nothing turns red.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_9"></a>|@ralphm commented|
|-|-|

exarkun beat me, so here some additional small comments:

 * There is spurious white space [here](http://twistedmatrix.com/trac/browser/branches/conch-tap-fix-3871/twisted/conch/test/test_tap.py#L8).
 * It would be nice to have a comment on how PAM is actually detected to be available. I get confused by secondary `ImportError`s.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve removed owner|
|-|-|

Everything handled, I think.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_11"></a>|@mwhudson commented|
|-|-|

Not a review, but this line:

+        L{ta.makeService} returns a L{TCPServer} instance, linked to

contains a typo.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

Thanks, fixed!

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_13"></a>|@mwhudson commented|
|-|-|

A few comments (entirely about the test, it seems):

 1. Is test_makeServicePamAuth a good enough test for the problem as described (i.e. 'twistd conch' doesn't work)?  Please at least add another test that monkey patches tap.pmauth to None.
 1. "L{tap.makeService} returns a L{TCPServer} instance, linked to L{OpenSSHFactory} whose" from the docstring is missing an article (probably "an", i.e. "L{tap.makeService} returns a L{TCPServer} instance, linked to **an** L{OpenSSHFactory} whose")
 1. test_makeServicePamAuth should use self.patch, not doing that stuff by hand.
 1. A comment explaining why the monkey-patching is there would be a definite bonus.
 1. The verification part of the test is testing considerably more than the docstring of the test method suggests, for example that the port is 22 and testing the entire set of checkers rather than just that the PAM one is there.  Could you break this up into a few testcases instead?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve removed owner|
|-|-|

Thanks a lot for the review, everything handled in r27091 I think.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @therve|
|-|-|

`twisted.conch.test.test_tap.MakeServiceTest.test_checkersPamAuth` fails.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

Here's the failure:

```
===============================================================================
[ERROR]: twisted.conch.test.test_tap.MakeServiceTest.test_checkersPamAuth

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/conch/test/test_tap.py", line 49, in test_checkersPamAuth
    service = tap.makeService(config)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/conch/tap.py", line 42, in makeService
    t.portal.registerChecker(PluggableAuthenticationModulesChecker())
exceptions.NameError: global name 'PluggableAuthenticationModulesChecker' is not defined
-------------------------------------------------------------------------------
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve removed owner|
|-|-|

Hum, sorry about that, I didn't test without PyPAM installed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set owner to @therve|
|-|-|

Windows had a test failure:

```
===============================================================================
[ERROR]: twisted.conch.test.test_tap

Traceback (most recent call last):
  File "c:\buildbot\twisted\WXP32-full2.4-scmikes-select\Twisted\twisted\trial\runner.py", line 552, in loadPackage
    module = modinfo.load()
  File "c:\buildbot\twisted\WXP32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 380, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "c:\buildbot\twisted\WXP32-full2.4-scmikes-select\Twisted\twisted\python\reflect.py", line 462, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "c:\buildbot\twisted\WXP32-full2.4-scmikes-select\Twisted\twisted\conch\test\test_tap.py", line 8, in ?
    from twisted.conch import tap
  File "c:\buildbot\twisted\WXP32-full2.4-scmikes-select\Twisted\twisted\conch\tap.py", line 9, in ?
    from twisted.conch import checkers, unix
  File "c:\buildbot\twisted\WXP32-full2.4-scmikes-select\Twisted\twisted\conch\unix.py", line 17, in ?
    import fcntl, tty
exceptions.ImportError: No module named fcntl
-------------------------------------------------------------------------------
```

:/

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve commented|
|-|-|

(In [27093]) Branching to 'conch-tap-fix-3871-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve removed owner|
|-|-|

Sigh. It should be ok.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set owner to @therve|
|-|-|

Seems to be working (not really, but working enough for this ticket).  Shame there are so many Conch cross-platform stumbling blocks. :(  Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_24"></a>|@therve set status to `closed`|
|-|-|

(In [27097]) Merge conch-tap-fix-3871-2

Author: therve
Reviewers: exarkun, mwh
Fixes [#3871](https://github.com/twisted/twisted/issues/3871)

Fixes a bad import in twisted.conch.tap preventing from running it at all
and add some test coverage for it.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
